src/java/org/apache/cassandra/index/sai/disk/RAMStringIndexer.java

```diff
@@ -0,0 +1,163 @@
+    private int[] lastSegmentRowID = new int[RAMPostingSlices.DEFAULT_TERM_DICT_SIZE];
+
+    public RAMStringIndexer()
+    {
+        bytesUsed = Counter.newCounter();
+
+        ByteBlockPool termsPool = new ByteBlockPool(new ByteBlockPool.DirectTrackingAllocator(bytesUsed));
+
+        termsHash = new BytesRefHash(termsPool);
+
+        slices = new RAMPostingSlices(bytesUsed);
+    }
+
+    public long estimatedBytesUsed()
+    {
+        return bytesUsed.get();
+    }
+
+    public boolean isEmpty()
+    {
+        return rowCount == 0;
+    }
+
+    /**
+     * EXPENSIVE OPERATION due to sorting the terms, only call once.
+     */
+    // TODO: assert or throw and exception if getTermsWithPostings is called > 1
+    public TermsIterator getTermsWithPostings()
+    {
+        final int[] sortedTermIDs = termsHash.sort();
+
+        final int valueCount = termsHash.size();
+        final ByteSliceReader sliceReader = new ByteSliceReader();
+
+        return new TermsIterator()
+        {
+            private int ordUpto = 0;
+            private final BytesRef br = new BytesRef();
+
+            @Override
+            public ByteBuffer getMinTerm()
+            {
+                BytesRef term = new BytesRef();
+                int minTermID = sortedTermIDs[0];
+                termsHash.get(minTermID, term);
+                return ByteBuffer.wrap(term.bytes, term.offset, term.length);
+            }
+
+            @Override
+            public ByteBuffer getMaxTerm()
+            {
+                BytesRef term = new BytesRef();
+                int maxTermID = sortedTermIDs[valueCount-1];
+                termsHash.get(maxTermID, term);
+                return ByteBuffer.wrap(term.bytes, term.offset, term.length);
+            }
+
+            public void close() {}
+
+            @Override
+            public PostingList postings()
+            {
+                int termID = sortedTermIDs[ordUpto - 1];
+                return slices.postingList(termID, sliceReader);
+            }
+
+            @Override
+            public boolean hasNext() {
+                return ordUpto < valueCount;
+            }
+
+            @Override
+            public ByteComparable next()
+            {
+                if (!hasNext())
+                    throw new NoSuchElementException();
+
+                termsHash.get(sortedTermIDs[ordUpto], br);
+                ordUpto++;
+                return asByteComparable(br.bytes, br.offset, br.length);
+            }
+
+            private ByteComparable asByteComparable(byte[] bytes, int offset, int length)
+            {
+                return v -> ByteSource.fixedLength(bytes, offset, length);
+            }
+        };
+    }
+
+    public long add(BytesRef term, int segmentRowId)
+    {
+        long startBytes = estimatedBytesUsed();
+        int termID = termsHash.add(term);
+
+        if (termID >= 0)
+        {
+            // firs time seeing this term, create the term's first slice !
```

Comment:
```suggestion
            // first time seeing this term, so create its first slice
```
It's not _that_ exciting